text/0000-roadmap-2017.md

```diff
@@ -0,0 +1,577 @@
+[futures library announcement]: http://aturon.github.io/blog/2016/08/11/futures/
+
+### Rust should integrate easily into large build systems
+
+When working with larger organizations interested in using Rust, one of the
+first hurdles we tend to run into is fitting into an existing build
+system. We've been exploring a number of different approaches, each of which
+ends up using Cargo (and sometimes `rustc`) in different ways, with different
+stories about how to incorporate crates from the broader crates.io ecosystem.
+Part of the issue seems to be a perceived overlap between functionality in Cargo
+(and its notion of compilation unit) and in ambient build systems, but we have
+yet to truly get to the bottom of the issues—and it may be that the problem is
+one of communication, rather than of some technical gap.
+
+By the end of 2017, this kind of integration should be *easy*: as a community,
+we should have a strong understanding of best practices, and potentially build
+tooling in support of those practices. And of course, we want to approach this
+goal with Rust's values in mind, ensuring that first-class access to the
+crates.io ecosystem is a cornerstone of our eventual story.
+
+### Rust should integrate easily with C++ code
+
+Rust's current support for interfacing with C is fairly strong, but wrapping a C
+library still involves tedious work mirroring declarations and writing C shims
+or other glue code. Moreover, many projects that are ripe for Rust integration
+are currently using C++, and interfacing with those effectively requires
+maintaining an alternative C wrapper for the C++ APIs. This is a problem both
+for Rust code that wants to employ existing libraries and for those who want to
+integrate Rust into existing C/C++ codebases.
+
+Why C++ rather than the myriad other languages we might focus on? Of any of the
+possible language integrations, C++ came up most consistently across both the
+general survey and the commercial outreach. And that's not too surprising: Rust
+is well-positioned as a C++ replacement, so it's natural for people with
+existing C++ codebases to be looking at Rust but needing a good integration
+story for moving incrementally. In addition, many of the features needed for
+high quality C/C++ binding support are prerequisites for high quality binding
+support in other languages, since runtime internals will generally be accessible
+through a C or C++ interface. So in general, C++ seems like the wisest place to
+focus on the language integration front for the moment.
+
+**The goal should be that using a C++ library in Rust is not much harder than
+using it in C++**. In other words, it should be possible to directly include C++
+headers (e.g., include! {myproject.hpp}) and have the extern declarations, glue
+code, and so forth get generated automatically. This means (eventually) full
+support for interfacing with C++ code that uses features like templates,
+overloading, classes and virtual calls, and so forth.
```

Comment:
Hmm. The text may be misleading in its emphasis. Let me explain my POV and see what people think. I think @jonathandturner may have a slightly different take, though I think we definitely agree on the end-points we're aiming for; @wycats also has a lot of well-informed opinions here. =)

In my opinion, both C and C++ are very important. I think we should emphasize steps that benefit *both* languages (I'll sketch out a possible plan below). But basically C++ is the only language that is not C that I think makes sense for the Rust teams themselves to pursue. C is certainly the "lingua franca", but there are plenty of big projects and libraries that use C++ too (off the top of my head: V8, LLVM if you want the full thing, Gecko). Anecdotally, I certainly get a lot of questions about C++ integration.

(To be clear, I definitely hope we will see work on integration with other languages, like Python etc. I just think that is best handled by other projects -- though if we find features that are needed, it'd be great to surface those as RFCs.)

Regarding C++ integration, I agree we're not going to implement a C++ compiler in the Rust compiler, and I don't think the roadmap should commit us to any such thing! =) What I do think is that we should work on reducing FFI boilerplate as much as we can. I think this starts just by taking a tool like bindgen and making it drop-dead simple to invoke (`include! { ... }`), perhaps leveraging progress on macros-1.1-style integration. 

Once we have that, we can extend this `include!` macro to start filling in the gaps in our existing C integration. So for example dealing with macros properly is a big pain right now. Projects like Ruby using macros as inline functions, basically, so to properly bind to those you have to generate C shims that your Rust code can call. But wouldn't it be nice if `include!` could do this for you? (You'd probably have to direct it some, e.g. to tell it types involved.)

Finally, we can start to leverage some of the awesome work that @emilio has been doing in Servo project for supporting C++. He can speak more to the details, but from what I understand he has a branch of bindgen that can create layout-compatible Rust structs for the C++ used in Gecko, which is no mean feat. What it doesn't support are things like virtual calls and so forth (it just leaves opaque vtables in the Rust structs, as I understand).

Now the C++ ABI is famously undefined and we don't want to get too deeply involved there. But this looks (to me) like a place we can leverage shims. We can basically generate `extern "C"` wrappers for C++ things you'd like to call. Using said shims, at least initially, will probably still be tedious (e.g., instead of calling `widget->draw()`, you call `widget_draw(widget)`) but at least you don't have to write the shims by hand. Maybe you initially also specify template instantiations by hand.

If we got those steps, we'd be in a much better shape. You can now generate boilerplate code easily. From there, there seem to be many directions we could go:

- generating more "ergonomic" wrappers, e.g. using Rust methods so you can call `widget.draw()`;
- improving efficiency by compiling the shims (using clang) to LLVM and then inlining them into Rust code;
- integrate the bindgen tool more deeply with the compiler, so that the compiler can request template instantiations as it monomorphizes, for example, or generate shims on the fly that use the C++ compiler to resolve overloading.

Anyway, this is getting afield, and obviously nothing in the roadmap commits us to any part of this path. But I wanted to sketch out an incremental vision that nonetheless gets us to relatively easy and feature-complete state, while remaining fairly agnostic in the compiler itself.